Lib/xml/etree/ElementTree.py

```diff
@@ -731,6 +731,8 @@ def write(self, file_or_filename,
         with _get_writer(file_or_filename, encoding) as (write, declared_encoding):
             if method == "xml" and (xml_declaration or
                     (xml_declaration is None and
+                     not (encoding.lower() == "unicode" and
+                          hasattr(file_or_filename, "write")) and
                      declared_encoding.lower() not in ("utf-8", "us-ascii"))):
```

Comment:
Why not just this?

`xml_declaration is None and encoding.lower() not in ("utf-8", "us-ascii", "unicode")`

Reply:
Because it will write an invalid XML file when pass a file name on non-UTF-8 locale. Run `test_write_to_filename_as_unicode` on non-UTF-8 locale.

When the default encoding in `open()` be UTF-8, and only UTF-8, and nothing except UTF-8, this code could be simplified.

It can even be done earlier if open the file in UTF-8 instead of the default encoding with `encoding='unicode'`. But such change is not for backporting. I'll create a separate PR for the main branch only.